lib/declare_schema/model/field_spec.rb

```diff
@@ -47,11 +47,9 @@ def round_up_mysql_text_limit(limit)
       end
 
       def initialize(model, name, type, position: 0, **options)
-        # TODO: TECH-5116
-        # Invoca change - searching for the primary key was causing an additional database read on every model load.  Assume
-        # "id" which works for invoca.
-        # raise ArgumentError, "you cannot provide a field spec for the primary key" if name == model.primary_key
-        name == "id" and raise ArgumentError, "you cannot provide a field spec for the primary key"
+        defined_primary_key = model.defined_primary_key
+
+        name.to_s == defined_primary_key and raise ArgumentError, "you may not provide a field spec for the primary key #{name.inspect}"
```

Comment:
There is a subtlety here that I think needs to be addressed with this check that has to do with how the `create_table` is generated. Primarily create table has the [following contract](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L154-L296). The key part being the slight difference in defining the primary key for the table. You can do one of the following two things:

1. Use `id` as the primary key and define the specific type to use in the `create_table` statement directly (we are doing this already to make `id` a `:bigint`
```ruby
create_table(:some_table, id: :bigint) do; end

CREATE TABLE some_table (
    id bigint DEFAULT NULL auto_increment PRIMARY KEY
)
```

2. Use a different column as the primary key and let the default rails type take over (in 4.2 this is int(11))
```ruby
create_table(:some_table, primary_key: 'guid') do; end

CREATE TABLE suppliers (
    guid int(11) DEFAULT NULL auto_increment PRIMARY KEY
)
```

3. Use a different column(s) as the primary key which are defined within the field specs for usage of custom type
```ruby
create_table(:some_table, primary_key: 'name') do |t|
  t.string :name
end

CREATE TABLE suppliers (
    name string DEFAULT NULL PRIMARY KEY
)
```

I think we need to make a choice right now to either support all, or block one/two of these until we get the support added.

Reply:
I agree with that background, but I'm not sure we have to do anything right now. This code is equivalent to what we've always had, right?

Reply:
(BTW it's an eternal headache in this code and SQL itself that primary keys are columns and also indexes. 99% of the time a single column, but 1% they are compound. So sometimes the PK is thought of as an attribute of the table, sometimes it's a boolean option on a column, and sometimes it's a special index. It would be great to standardize on one approach. The middle option there doesn't work for compound PKs--because you need ordering too--so it's not a great choice. The latter is probably the most general and flexible. In `declare_schema` we do that some, with a convention of the index name being PRIMARY.)

Reply:
True the contract isn't changing. This would be an enhancement so it's not necessary right now.